exp/internal/controllers/machinepool_controller.go

```diff
@@ -253,33 +253,51 @@ func (r *MachinePoolReconciler) reconcile(ctx context.Context, cluster *clusterv
 	return res, kerrors.NewAggregate(errs)
 }
 
-func (r *MachinePoolReconciler) reconcileDelete(ctx context.Context, cluster *clusterv1.Cluster, mp *expv1.MachinePool) error {
-	if ok, err := r.reconcileDeleteExternal(ctx, mp); !ok || err != nil {
-		// Return early and don't remove the finalizer if we got an error or
-		// the external reconciliation deletion isn't ready.
-		return err
-	}
+func (r *MachinePoolReconciler) reconcileDelete(ctx context.Context, cluster *clusterv1.Cluster, machinepool *expv1.MachinePool) error {
+	deleteAllowed, clusterClient, err := r.deleteMachinePoolAllowed(ctx, cluster, machinepool)
+
+	// Check for cluster allowing delete or machinePool delete timeout.
+	if deleteAllowed || r.isMachinePoolNodesDeleteTimeoutPassed(machinepool) {
+		if ok, err := r.reconcileDeleteExternal(ctx, machinepool); !ok || err != nil {
```

Comment:
Why is reconcileDeleteExternal within this if block?
If there node timeout is infinite and clusterClient fails we block deletion for ever, is that intentional?
Can we add code comment elaborating the why for this logic? Also additional test cases with the scenarios this is trying to account for might help.

Reply:
wouldn't this all we need to achieve the goal don't hang forever on Node deletion?
```
	if ok, err := r.reconcileDeleteExternal(ctx, mp); !ok || err != nil {
		// Return early and don't remove the finalizer if we got an error or
		// the external reconciliation deletion isn't ready.
		return err
	}

	if !NodeDeletionTimeoutMet() {
		if err := r.reconcileDeleteNodes(ctx, cluster, mp); err != nil {
			// Return early and don't remove the finalizer if we got an error.
			return err
		}
	}

	controllerutil.RemoveFinalizer(mp, expv1.MachinePoolFinalizer)
	return nil
```

Reply:
well, the idea was to tie the machinePool resources (InfrastructureRef) with the delete timeout define in the machineSpec.

Reply:
>well, the idea was to tie the machinePool resources (InfrastructureRef) with the delete timeout define in the machineSpec.

Why? the deletion timeout is meant for the Node kube resource. The infra must be deleted before the node, otherwise stateful pods primitives are in risk of being corrupted. Then we delete the Node to signal the kas and we have the timeout in case something catastrophic happened e.g. kas lb is broken we don't block forever.

There's no timeout primitive for infra deletion today. If we think that might be needed is a different discussion we can have separately but nothing articulated so far makes me think why we'd need that.
Thanks for iterating with me!